File: src-tauri/src/bluetooth/device_data.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use service::ServiceData;

use super::connected_device::ConnectedDevice;

pub mod characteristic;
pub mod service;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DeviceData {
    pub id: String,
    pub name: Option<String>,
    pub address: Option<String>,
    pub manufacturer_data: Option<HashMap<u16, Vec<u8>>>,
    pub services: Vec<ServiceData>,
}

impl From<&ConnectedDevice> for DeviceData {
    fn from(device: &ConnectedDevice) -> Self {
        let services = device.services.values().map(|c| c.into()).collect();

        Self {
            id: device.device.id.to_string(),
            name: device.device.name.clone(),
            address: device.device.address.clone(),
            manufacturer_data: device.device.manufacturer_data.clone(),
            services,
        }
    }
}

File: src/app/device-details/device-data.ts
export interface CharacteristicData {
  uuid: string;
  read: boolean;
  write: boolean;
  write_without_response: boolean;
  notify: boolean;
}

export interface ServiceData {
  uuid: string;
  characteristics: CharacteristicData[];
}

export interface DeviceData {
  id: string;
  name: string;
  address: string;
  manufacturer_data: Record<string, any>;
  services: ServiceData[];
}
